### DISCUSSION TITLE
How to properly add catch2 (or any other dev only library)

### DISCUSSION THREAD

**Discussion by fpicchi** [friendlyanon/cmake-init#8] [Q&A] [answered]
I've been trying to understand what's the proper way to add catch2 dependency to my project.
I've seen that the generated cmakes have a separated "dev"mode, but still I'm a bit lost.
Regards,
Federico

P.S. to exclude tests from cppcheck is it better to change the cppcheck preset, or to append the -itest flag to dev preset?

**Comment by friendlyanon** [comment, marked as answer]
The code you have to include in the `test/CMakeLists.txt` file looks like this:

```diff
@@ -10,11 +10,8 @@
   enable_testing()
 endif()

+find_package(Catch2 REQUIRED)
+include(Catch)
+
 add_executable(example_test source/example_test.cpp)
-target_link_libraries(example_test PRIVATE example::example)
+target_link_libraries(example_test PRIVATE example::example Catch2::Catch2)
 target_compile_features(example_test PRIVATE cxx_std_17)

-add_test(NAME example_test COMMAND example_test)
+catch_discover_tests(example_test)
```

This consists of two parts:

### Instruct CMake to go looking for a `Catch2` package

`find_package` operates in two modes, module mode and config mode, but in most cases you want to only care about the config mode. For config mode, there is a [search procedure](https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure) for finding packages. It's just a long long list of paths CMake has to search, it's nothing complicated.

If Catch2 is installed to your system by your system package manger, then CMake will be able to find it using just the default prefixes that it searches. If you are using vcpkg, then your `find_package` call is already doing a bit more and this is taken care for you. If you use conan, then you have to use the `cmake_find_package_multi` generator and pass the install path to CMake:

```sh
conan install catch2/2.13.6@ -if deps
# Windows
cmake -S . -B build -D "CMAKE_PREFIX_PATH=%cd:\=/%/deps"
# Unix based
cmake -S . -B build -D "CMAKE_PREFIX_PATH=$PWD/deps"
```

Generally speaking, `CMAKE_PREFIX_PATH` is the list variable that you want to pass from the CLI to add new prefixes to search in the search procedure. This variable is what you use to pass manually installed dependencies as well.

### Using the result of the previous command for Catch2's CMake integration

Catch2 comes with a handy [CMake integration](https://github.com/catchorg/Catch2/blob/devel/docs/cmake-integration.md), which is available after finding it.

**Reply by fpicchi** [reply]
Thank you a lot.

**Comment by friendlyanon** [comment]
For the cppcheck part, you should include that in the `cppcheck` preset. When separating arguments, you have to use semicolons, similar to how it's done in the `clang-tidy` preset.

**Comment by friendlyanon** [comment]
Catch2 is now provided by default when a package manager is used with the generated project. This is a not-so-recent addition.
